modyn/trainer_server/internal/dataset/per_class_online_dataset.py

```diff
@@ -0,0 +1,94 @@
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        dataset_id: str,
+        bytes_parser: str,
+        serialized_transforms: list[str],
+        storage_address: str,
+        selector_address: str,
+        training_id: int,
+    ):
+        super().__init__(
+            pipeline_id,
+            trigger_id,
+            dataset_id,
+            bytes_parser,
+            serialized_transforms,
+            storage_address,
+            selector_address,
+            training_id,
+        )
+
+        self._filtered_label = None
+
+    # pylint: disable=too-many-locals, too-many-branches
+    def __iter__(self) -> Generator:
+        assert self._filtered_label is not None
+        worker_info = get_worker_info()
+        if worker_info is None:
+            # Non-multithreaded data loading. We use worker_id 0.
+            worker_id = 0
+        else:
+            worker_id = worker_info.id
+
+        if self._first_call:
+            self._first_call = False
+            self._debug("This is the first run of iter, making gRPC connections.", worker_id)
+            # We have to initialize transformations and gRPC connections here to do it per dataloader worker,
+            # otherwise the transformations/gRPC connections cannot be pickled for the new processes.
+            self._init_transforms()
+            self._init_grpc()
+            self._key_source.init_worker()
+            self._uses_weights = self._key_source.uses_weights()
+            self._silence_pil()
+            self._debug("gRPC initialized.", worker_id)
+
+        assert self._transform is not None
+        self._num_partitions = self._key_source.get_num_data_partitions()
+        self._info(f"Total number of partitions will be {self._num_partitions}", worker_id)
+
+        keys, data, labels, weights = self._get_data(worker_id=worker_id, partition_id=0)
+
+        for partition in range(self._num_partitions):
+            num_samples_on_this_partition = len(keys)
+            # We (arbitrarily) fetch the next partition when we have seen 80% of the current partition
+            fetch_next_partition_idx = int(num_samples_on_this_partition * 0.8)
+            self._info(f"Train on partition {partition}, on {num_samples_on_this_partition} batches", worker_id)
+
+            for idx, data_tuple in self._get_data_iterator(keys, data, labels, weights):
+                key, sample, label, weight = self._unpack_data_tuple(data_tuple)
+
+                if partition < self._num_partitions - 1 and idx == fetch_next_partition_idx:
+                    # TODO(#175) in case this blocks training
+                    new_keys, new_data, new_labels, new_weights = self._get_data(
+                        worker_id=worker_id, partition_id=partition + 1
+                    )
+                if label == self._filtered_label:
+                    yield self._yield_samples(key, sample, label, weight)
```

Comment:
hmm alternatively, couldnt we use the regular onlinedataset, and give it the optional parameter filtered label, and if it is not none, then we do this check? then we avoid introducing a new class. anything speaks against that?

Reply:
In my opinion, it is not worth changing the online dataset for a temporary feature. It is already complex enough, I would avoid adding to it.

I have now changed the iteration loop to 
```
data_tuple = self._get_data_tuple(key, sample, label, weight)
if data_tuple is not None:
   yield data_tuple
```

So we just have to overload the method `_get_data_tuple` and filter there

Reply:
So basically this is the only difference
```
def _get_data_tuple(self, key: int, sample: bytes, label: int, weight: Optional[float]) -> Optional[Tuple]:
    assert self.filtered_label is not None

    if self.filtered_label != label:
        return None
    return super()._get_data_tuple(key, sample, label, weight)
```